### PR TITLE
FlaxGPTNeo

### PR DESCRIPTION
**Pull request by patil-suraj**
# What does this PR do?

This PR adds the Flax version of GPTNeo. For local attention, it uses the fix proposed by @finetuneanon in #11630.

Thanks a lot, @finetuneanon for proposing the solution, it's especially important in JAX/Flax where we can't have dynamic shapes.

Official GPTNeo flax checkpoints are up on the hub and slow tests are passing.

### CODE REVIEW

File: src/transformers/models/gpt_neo/modeling_flax_gpt_neo.py [hunk 1]
```diff
@@ -0,0 +1,671 @@
...
+            dtype=self.dtype,
+            kernel_init=jax.nn.initializers.normal(self.config.initializer_range, self.dtype),
+        )
+
+        self.q_proj, self.k_proj, self.v_proj = dense(use_bias=False), dense(use_bias=False), dense(use_bias=False)
+        self.out_proj = dense()
+
+        self.causal_mask = make_causal_mask(jnp.ones((1, config.max_position_embeddings), dtype="bool"), dtype="bool")
+        if self.attention_type == "local":
+            self.causal_mask = self.causal_mask ^ jnp.tril(self.causal_mask, -config.window_size)
```
Comment (patrickvonplaten) [inline src/transformers/models/gpt_neo/modeling_flax_gpt_neo.py:130]: nice!

Comment (patrickvonplaten) [reply src/transformers/models/gpt_neo/modeling_flax_gpt_neo.py:130]: maybe an additional comment here would be nice

**Review by patrickvonplaten — APPROVED** [review]
Awesome - very clean!

File: tests/test_modeling_flax_gpt_neo.py [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+
+                batch_size, seq_length = pt_inputs["input_ids"].shape
+                rnd_start_indices = np.random.randint(0, seq_length - 1, size=(batch_size,))
+                for batch_idx, start_index in enumerate(rnd_start_indices):
+                    pt_inputs["attention_mask"][batch_idx, :start_index] = 0
+                    pt_inputs["attention_mask"][batch_idx, start_index:] = 1
+                    prepared_inputs_dict["attention_mask"][batch_idx, :start_index] = 0
+                    prepared_inputs_dict["attention_mask"][batch_idx, start_index:] = 1
+                pt_model = pt_model_class(config).eval()
+                fx_model = model_class(config, dtype=jnp.float32)
```
Comment (LysandreJik) [inline tests/test_modeling_flax_gpt_neo.py:252]: Is `fx_model` a common name for Flax models? It reminds of torch.fx

**Review by LysandreJik — APPROVED** [review]
This is very nice, thanks for working on it @patil-suraj!

File: tests/test_modeling_flax_gpt_neo.py (hunk 2 above)
Comment (patil-suraj) [reply tests/test_modeling_flax_gpt_neo.py:252]: Aah, yeah this is confusing. Maybe we could use `flx` or just `flax` for flax models. (cc @patrickvonplaten )

**Pull request merged**